### DISCUSSION TITLE
Client side for root renant and for regular tenants - two wasm projects?

### DISCUSSION THREAD

**Discussion by ThomasVague** [fullstackhero/blazor-wasm-boilerplate#132] [Q&A]
Hi, I just want to start of saying these projects are awesome!

So I have a question regarding having different clients consuming the web api. 
My scenario : 
The project is for a company selling construction services to different construction companies. Delivering services like project managing for building houses, registering work hours etc.
So the idea is to have an admin user for each construction company (tenant), and root admin users who can create new tenants, manage licenses purchased for each tenant etc. In general just administer the tenants. Much like the wasm project is setup at the moment.

However, I don't want the admin user or any basic user for that matter in each tenant to have access to features like hangfire etc. So in that regard, the client side for a root tenant would be vastly different than for a "regular" tenant.  The client side for a regular tenant would be services like 
"create new project"
"register hours"
"create a checklist", etc...

whilst the admin client side would be mostly if not entirely to administer tenants etc..

So I currently see 3 different options:
1. Two distinct solutions - one for root tenant administration and one for a basic tenant with seperate wasm hosts.

2. One solution, but with two wasm projects. And then the host will serve a wasm client depending on the url, So something like  companysite.com for a regular tenant and companysite.com/admin or admin.companysite.com for root tenant.

3. Have one solution with one wasm client and configure the client to show site features depending on roles, rights, authorization etc.

Any thoughts on this?

**Comment by fretje** [comment]
That's an interesting topic for sure.

The way FSH is set up now, there's a special kind of tenant, the "root" tenant, which can also "administer tenants".

I personally think the "tenant administration" functionality should probably be better of to be split into a different application. with its own authentication/authorization (where only one or a few users have access) , and it should only be concerned about creating and administering new tenants (and creating/managing the databases in the backend). It wouldn't have anything application specific in it.

The application itself would then have nothing tenant related in it.

That way, the concept of "root" tenant would go away. Every tenant would be a regular tenant, and it might make things simpler I think.

Not sure though... Still kind of processing... got me thinking for sure ;-)
